onnxruntime/core/graph/function_impl.h

```diff
@@ -27,17 +27,24 @@ class FunctionImpl final : public Function {
   // a Function Op. This takes in a FunctionProto and constructs function body
   // from it. The function body initialization happens during model load in graph resolve
   // phase.
-  FunctionImpl(const onnxruntime::Graph& graph,
+  // model_local_functions contains domain:optype to model_local_functions map. This is
+  // used to resolve and initialize nested functions.
+  FunctionImpl(onnxruntime::Graph& graph,
                const onnxruntime::NodeIndex& node_index,
                const ONNX_NAMESPACE::FunctionProto& onnx_func,
-               const logging::Logger& logger);
+               const std::unordered_map<std::string, const ONNX_NAMESPACE::FunctionProto*>& in_model_function_protos,
+               std::vector<std::unique_ptr<onnxruntime::Function>>& function_container,
+               const logging::Logger& logger,
+               bool isNestedFunction = false);
```

Comment:
It would help to add a couple of lines of explanatory documentation. First, a single ONNX model-local function F can be called from both the top-level (main) graph as well as from other model-local functions. Thus, it is a bit ambiguous in the ONNX context to classify a function itself as nested or non-nested. I guess here we use a FunctionImpl to represent a specific call to a function ... so if we have multiple calls to F in a model, there will be multiple FunctionImpls (one for each call)? If this is the case, we could just add a couple of lines of comments to clarify this.

Reply:
It would also help to add comments to explain the current design in terms of when what happens: (a) When the FunctionImpl are created, and (b) When type-inference happens, especially if it is different for nested and non-nested functions.

Reply:
>> so if we have multiple calls to F in a model, there will be multiple FunctionImpls (one for each call)?
Yes, in current approach we create FuncImp for every occurrence of a function. We can share the FuncImpl for context INDEPENDENT functions. However, I think the ideal optimization here is to move to lazy initialization. 

I will add more comments